Covert to async with tokio + hyper crates for http logics

ZhidongPeng opened the pull request:
- Convert multiple threads proxy_listener to async mode proxy_server 
- replace custom http modules with hyper crates
- Limit the normal proxied request body size to 100KB

ZhidongPeng committed:
Covert to async with tokio + hyper crates for http logics (#8)

Convert multiple threads proxy_listener to async mode proxy_server
replace custom http modules with hyper crates

ZhidongPeng committed:
skip UT for bpf files

ZhidongPeng committed:
Merge branch 'dev' into dev1

ZhidongPeng committed:
Update UT

ZhidongPeng committed:
Merge branch 'dev1' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

ZhidongPeng committed:
Merge branch 'dev' into dev1

ZhidongPeng committed:
Merge branch 'dev' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

ZhidongPeng committed:
Merge branch 'dev1' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

ZhidongPeng committed:
Merge branch 'dev' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

vmarcella commented on proxy_agent/Cargo.toml:
```diff
@@ -18,6 +18,14 @@ url = "2.3.1"                 # parse url string
+hyper-util = { version = "0.1", features = ["tokio"] }
+http = "1.1.0"
+## reqwest doesnot build for x86_64-unknown-linux-musl
+# reqwest = "0.12.5"            # http client
```
Can we just get rid of reqwest entirely from our `Cargo.toml` now that we're including `hyper`?

ZhidongPeng replied:
Let's mark it why we do not use reqwest at this moment.

jeremycline replied:
I don't think leaving it here, commented out, makes sense. Put something in an issue tracker for this if you plan to eventually use it, pointing to some upstream issue for whatever the build problem is. As it stands, I'm left with more questions than answers.

ZhidongPeng replied:
removed.

vmarcella commented on proxy_agent/src/host_clients/imds_client.rs:
```diff
@@ -14,49 +14,27 @@ pub struct ImdsClient {
-    pub fn new(ip: &str, port: u16, shared_state: Arc<Mutex<SharedState>>) -> Self {
+    pub fn new(ip: String, port: u16, shared_state: Arc<Mutex<SharedState>>) -> Self {
         ImdsClient {
             ip: ip.to_string(),
```
Do we need this `to_string()` call now that ip is being passed in as a string?

ZhidongPeng replied:
good catch.

vmarcella commented on proxy_agent/src/host_clients/imds_client.rs:
```diff
@@ -14,49 +14,27 @@ pub struct ImdsClient {
-    pub fn get_imds_instance_info(&self) -> std::io::Result<InstanceInfo> {
-        const IMDS_URI: &str = "/metadata/instance?api-version=2018-02-01";
+    pub async fn get_imds_instance_info(&self) -> std::io::Result<InstanceInfo> {
+        const IMDS_URI: &str = "metadata/instance?api-version=2018-02-01";
```
I'm not too sure what the actual impact of declaring this const inside the function, but can we define this constant at the top of the module or within `common/constants.rs`? I'm sure the Rust compiler will optimize this to not be redeclared every function call, but I think because it's const we should just move this outside of the function scope.

vmarcella commented on proxy_agent/src/host_clients/wire_server_client.rs:
```diff
@@ -17,153 +13,96 @@ pub struct WireServerClient {
-    pub fn new(ip: &str, port: u16, shared_state: Arc<Mutex<SharedState>>) -> Self {
+    pub fn new(ip: String, port: u16, shared_state: Arc<Mutex<SharedState>>) -> Self {
         WireServerClient {
             ip: ip.to_string(),
```
Same as with the IMDS client, do we still need this `ip.to_string()` call now that ip is being passed in as a string?

vmarcella commented on proxy_agent/src/host_clients/wire_server_client.rs:
```diff
@@ -17,153 +13,96 @@ pub struct WireServerClient {
+            None,
+        )?;
+        let response = match http::send_request(
+            &self.ip.to_string(),
```
Same as the comment before with regards to ip already being a string

vmarcella commented on proxy_agent/src/key_keeper.rs:
```diff
@@ -97,18 +98,18 @@ fn poll_secure_channel_status(
 
-            let sleep =
-                if key_keeper_wrapper::get_current_secure_channel_state(shared_state.clone())
+            let sleep = if key_keeper_wrapper::get_current_secure_channel_state(shared_state.clone())
```
Would it be worth breaking out the sleep duration calculation into it's own function, or at the very least breaking the conditions into their own variables to make the comparison itself more descriptive?
```rust
#[inline]
pub fn get_sleep_interval(shared_state: Arc<Mutex<SharedState>>, fallback_interval: Duration) -> Duration {
    let in_unknown_state = key_keeper_wrapper::get_current_secure_channel_state(shared_state.clone()) == UNKOWN_STATE;
    let fallback_interval_is_longer = interval > FREQUENT_PULL_INTERVAL;
    let has_not_timed_out = helpers::get_elapsed_time_in_millisec() < FREQUENT_PULL_TIMEOUT_IN_MILLISECONDS;
    
    if in_unknown_state && fallback_interval_is_longer && has_not_timed_out {
        return FREQUENT_PULL_INTERVAL;
    }
    return interval;
}

ZhidongPeng replied:
it is the same thing since the if condition already breakout for multiple lines

vmarcella commented on proxy_agent/src/key_keeper.rs:
```diff
@@ -97,18 +98,18 @@ fn poll_secure_channel_status(
+            } else {
+                interval
+            };
             thread::sleep(sleep);
```
With the new concurrent model, does `thread::sleep()` prevent other things from happening inside the main thread? Should we use tokios builtin [sleep](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html) function instead, or does it not matter here?

ZhidongPeng replied:
good catch

vmarcella commented on proxy_agent/src/key_keeper/key.rs:
```diff
@@ -629,112 +627,88 @@ impl Clone for Key {
     Ok(status)
 }
 
 // base_url must end with '/'
```
Given that these functions expect `base_url` to end with '/', can we just add an assertion that checks if it does in the function call?

ZhidongPeng replied:
make sense, I use the url.join to generate the url string and caught the parse-errors.

vmarcella commented on proxy_agent/src/key_keeper/key.rs:
```diff
@@ -629,112 +627,88 @@ impl Clone for Key {
+    http::read_response_body(response).await
 }
 
 // base_url must end with '/'
```
Same as above

vmarcella commented on proxy_agent/src/proxy/proxy_connection.rs:
```diff
@@ -2,21 +2,41 @@
+impl ConnectionContext {
+    // try make sure the request could skip the sig
+    // and stream the body to the server directly
+    pub fn need_skip_sig(&self) -> bool {
```
NIT: Maybe we can rename this to `can_skip_sig` or `should_skip_sig`?

ZhidongPeng replied:
renamed to should_skip_sig.

vmarcella commented on proxy_agent/src/proxy/proxy_summary.rs:
```diff
@@ -6,6 +6,7 @@ use serde_derive::{Deserialize, Serialize};
 #[derive(Serialize, Deserialize)]
 #[allow(non_snake_case)]
 pub struct ProxySummary {
+    pub id: u128,
```
Is there any reason that the id for the Proxy Summary is so wide?

ZhidongPeng replied:
yeah, we need id when log the proxy summary, we could use this id to lookup the details log.

vmarcella commented on proxy_agent/src/redirector/windows.rs:
```diff
@@ -175,14 +173,14 @@ pub fn lookup_audit(
 }
 
-pub fn get_audit_from_redirect_context(tcp_stream: &TcpStream) -> std::io::Result<AuditEntry> {
+pub fn get_audit_from_redirect_context(raw_socket: usize) -> std::io::Result<AuditEntry> {
```
NIT: I'd rename this to `raw_socket_id` just to make it a bit more explicit, but I'm not hard pressed about it since you can infer that from it's `usize` type

vmarcella commented on proxy_agent/src/shared_state.rs:
```diff
@@ -108,13 +117,73 @@ impl Default for SharedState {
     }
 }
 
+pub mod shared_state_wrapper {
```
NIT: I think this module is better named `shared_state_helpers`, as a lot of the functions are helpers for properly working on the shared state object or accessing things from it.

ZhidongPeng replied:
we have the name pattern with wrapper

vmarcella commented on proxy_agent/src/service.rs:
```diff
@@ -32,22 +32,40 @@ pub fn start_service(shared_state: Arc<Mutex<SharedState>>) {
 }
 
+fn start_service_async(shared_state: Arc<Mutex<SharedState>>) {
+    _ = std::thread::Builder::new().spawn(move || {
```
Instead of building a new thread directly, should we let [tokio](https://tokio.rs/tokio/tutorial/spawning) handle that?

vmarcella commented on proxy_agent/src/shared_state.rs:
```diff
@@ -108,13 +117,73 @@ impl Default for SharedState {
+    ) -> std::io::Result<()> {
+        if shared_state
+            .lock()
+            .unwrap()
```
Will this cause a panic if the lock cannot be unwrapped/acquired?

ZhidongPeng replied:
this is by design; we keep all the static variables in this shared_state and use arc/mutex to guarantee the threads safe. if the unwrap failed, the application got into bad state, it has to be panic.

vmarcella commented on proxy_agent/src/telemetry/event_reader.rs:
```diff
@@ -55,17 +56,29 @@ pub fn start_async(
     _ = thread::Builder::new()
```
Same as a comment left earlier, should we use tokio to spawn this thread?

ZhidongPeng replied:
I make this function as async.

jeremycline replied:
I'd recommend using the `tokio::main` macro which will set up the runtime for the main thread. Then you can get rid of the entire `get_runtime` thing and just spawn the task.

ZhidongPeng committed:
resolve comments

ZhidongPeng committed:
Merge branch 'dev' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

vmarcella commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,394 @@
-    let mut response_without_body;
-    match read_response_without_body(&mut response_reader) {
-        Ok(r) => response_without_body = r,
+pub fn get_request(
```
Would it make sense to refactor this to `build_request`? At initial glance I thought this was for creating a `GET` request, but after looking at a few examples it was more clear to me that this was for building requests of any http method.

vmarcella reviewed (dismissed):
This iteration looks good to me, thanks for making everything else async! I'll give it an approval and wait for Jeremy to give it a further review.

ZhidongPeng committed:
rename to http::build_request

ZhidongPeng committed:
fix bug: forward response headers to client apps.

jeremycline commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,394 @@
-    }
-
-    Ok(request)
+pub fn htons(u: u16) -> u16 {
```
This is just a confusing (to those unfamiliar with C) rename of `to_be()` without documentation. I don't see any reason for this to exist.

ZhidongPeng replied:
This is particular function in networking/socket programming. Let me add bit comments to this function.

jeremycline replied:
Yes, I'm familiar with the C function. There's no reason to wrap a clear, well-documented function provided by the Rust standard library with a different one that only makes sense if you know C. It provides no value, but it does make readers have to do the translation in their head. Please remove it.

ZhidongPeng replied:
removed.

jeremycline commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,394 @@
-    response.set_body(receive_body_internal(&mut reader, content_length)?);
-
-    Ok(response)
+pub fn ntohs(u: u16) -> u16 {
```
Same note as above

jeremycline commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,394 @@
+                format!("Failed to send request to {}: {}", uri_str, e),
+            ))
         }
+    };
```
It's more idiomatic to use `map_err`, but as I've recommended before, you should just use the `anyhow` library.

ZhidongPeng replied:
used map_err

jeremycline commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,394 @@
             }
         };
+        if let Some(chunk) = frame.data_ref() {
+            body_string.push_str(&String::from_utf8_lossy(chunk));
```
Is `from_utf8_lossy` truly what you want? The response should include a content type header including the character set in use. If it's not UTF-8 it seems like that should be an error instead of a lossy conversion.

ZhidongPeng replied:
make sense.

jeremycline commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,394 @@
-                return Ok(data);
+pub async fn read_response_body<T>(
+    mut response: hyper::Response<hyper::body::Incoming>,
+    json_format: bool,
```
Wouldn't it be more correct to inspect the Response `Content-Type` header to determine if it's JSON?

ZhidongPeng replied:
read content_tye (json/xml), charset_type (utf-8/utf-16) from Content-Type header.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,694 @@
+    shared_state: Arc<Mutex<SharedState>>,
+) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
+    let (head, body) = request.into_parts();
+    let whole_body = match body.collect().await {
```
There doesn't appear to be any limit on the request size. Even though it looks like this happens after the authentication check, accepting arbitrarily large requests is probably not a good idea. There's a bunch of pre-built middleware for this kind of thing in [tower](https://tower-rs.github.io/tower/tower/) which should integrate nicely with Hyper (you can probably also implement authentication as a Tower service, as well).

ZhidongPeng replied:
Good idea to keep GPA service not OOM. Let me discuss with the host team to decide the size limitation. Temporally set limit to 100KB.

jeremycline commented on proxy_agent/src/redirector.rs:
```diff
@@ -45,14 +45,14 @@ impl AuditEntry {
+pub async fn start_async(local_port: u16, shared_state: Arc<Mutex<SharedState>>) {
+    tokio::spawn(async move {
+        start(local_port, shared_state).await;
     });
```
This function doesn't need to be marked `async` since it never needs to await anything. Additionally, there's no need to wrap the `start()` function in an async block as it's already returning a future you can pass to `tokio::spawn`:
```suggestion
pub fn start_async(local_port: u16, shared_state: Arc<Mutex<SharedState>>) {
    tokio::spawn(start(local_port, shared_state));
```

This applies to a number of other functions in this PR as well.

ZhidongPeng replied:
removed the "async move" within all tokio::spwan function.

jeremycline commented on proxy_agent/src/shared_state.rs:
```diff
@@ -72,6 +77,10 @@ impl SharedState {
         SharedState {
+            runtime: Some(Arc::new(Mutex::new(
+                tokio::runtime::Runtime::new().unwrap(),
+            ))),
```
I really wouldn't recommend this. Use the [tokio::main](https://docs.rs/tokio/latest/tokio/attr.main.html) macro.

ZhidongPeng replied:
switched to tokio::main and convert all proxy_agent to async mode.

jeremycline commented on proxy_agent/src/shared_state.rs:
```diff
@@ -108,13 +117,73 @@ impl Default for SharedState {
             telemetry_reader_shutdown: false,
             telemetry_logger_shutdown: false,
             telemetry_logger_status_message: UNKNOWN_STATUS_MESSAGE.to_string(),
+            mock_server_ip: None,   // test only
```
I'd recommend not putting test things here, but if you absolutely have to you can put them behind `#cfg(test)`

ZhidongPeng replied:
removed.

jeremycline commented on proxy_agent/src/telemetry/event_reader.rs:
```diff
@@ -447,8 +457,10 @@ mod tests {
-    #[test]
-    fn test_event_reader_thread() {
+    // this test is to test the event reader thread, it reads events from the events folder and send to wire server
+    // it requires more threads to run server and client
```
Do the tests hang or something with a single-thread runtime? They shouldn't, and if they do that's an indication that something's wrong with the code.

ZhidongPeng replied:
removed the "flavor = "multi_thread", worker_threads = 2"

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,694 @@
+            proxy_listener_wrapper::set_status_message(shared_state.clone(), message.to_string());
+            logger::write_warning(message.to_string());
+            return Ok(());
+        }
```
This is not really how the cancellation token is meant to be used. Instead, you can `select!` on the cancellation token along with the future you want to let someone else cancel.

ZhidongPeng replied:
I switched to Shutdown state in shared_state as before.

jeremycline reviewed (commented):
I didn't do a full review, but this has some notes with what I had time to get to today.

ZhidongPeng committed:
Merge branch 'dev' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

ZhidongPeng committed:
resolve comments (#10)

- Converted to tokio::main and switch all the thread::sleep
- removed thread_priorty dependency
- resolved all comments

ZhidongPeng committed:
exclude main and redirect rs files for cc

ZhidongPeng committed:
Merge branch 'dev' into dev1

jeremycline commented on proxy_agent/src/main.rs:
```diff
@@ -72,22 +86,47 @@ fn main() {
+    //     if let Err(e) = windows::run_service().await {
+    //         logger::write_error(format!("{e}"));
+    //     }
+    // });
```
Let's not check in commented-out code.

ZhidongPeng replied:
good catch, removed.

jeremycline commented on proxy_agent/src/main.rs:
```diff
@@ -72,22 +86,47 @@ fn main() {
+        }
+        None => {
+            logger::write_error("Failed to get the tokio runtime handle.".to_string());
+        }
```
I don't think it makes sense for the service to continue running in this case. This is a fatal programmer error since this can only happen if `ASYNC_RUNTIME_HANDLE` wasn't initialized prior to this function being invoked. Something like `let handle = ASYNC_RUNTIME_HANDLE.get().expect("You must provide the Tokio runtime handle before this function is called")` or something to that effect.

ZhidongPeng replied:
make sense, replaced with your suggestion.

jeremycline commented on proxy_agent/src/main.rs:
```diff
@@ -72,22 +86,47 @@ fn main() {
-        logger::write_error(format!("{e}"));
+    match ASYNC_RUNTIME_HANDLE.get() {
+        Some(handle) => {
+            handle.spawn(windows::run_service());
```
Is the Windows service lifetime tied to `proxy_agent_windows_service_main`? In other words, do you want this call to block until `run_service` completes? If so you want [block_on](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.block_on).

ZhidongPeng replied:
we need block_on

jeremycline commented on proxy_agent/src/monitor.rs:
```diff
@@ -8,26 +8,23 @@ use crate::{
+    tokio::spawn(async move {
+        start(interval, shared_state).await;
+    });
 }
```
```suggestion
pub fn start_async(interval: Duration, shared_state: Arc<Mutex<SharedState>>) {
    tokio::spawn(start(interval, shared_state));
}
```

ZhidongPeng replied:
removed this one-line wrapper fun

jeremycline commented on proxy_agent/src/provision.rs:
```diff
@@ -49,7 +49,7 @@ pub fn provision_timeup(provision_dir: Option<PathBuf>, shared_state: Arc<Mutex<
 }
 
-pub fn start_event_threads(shared_state: Arc<Mutex<SharedState>>) {
+pub async fn start_event_threads(shared_state: Arc<Mutex<SharedState>>) {
```
This function doesn't call any asynchronous functions. It doesn't need to be marked async as it never needs to await any results.

ZhidongPeng replied:
yes, it is no longer needed after previous changes.

jeremycline commented on proxy_agent/src/proxy/proxy_connection.rs:
```diff
@@ -2,21 +2,41 @@
     pub now: Instant,
-    pub cliams: Option<Claims>,
+    pub claims: Option<Claims>,
+    pub method: String,
```
Use the [hyper-provided type for this](https://docs.rs/hyper/latest/hyper/struct.Method.html)

ZhidongPeng replied:
switched to hyper::Method struct instead of String.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,689 @@
+    // Get the dst ip and port to remote server
+    let (ip, port);
+    ip = redirector::ip_to_string(entry.destination_ipv4);
+    port = u16::from_be(entry.destination_port); // convert a 16-bit number from network byte order to host byte order
```
Why doesn't whatever type `entry` is handle converting the port to the host byte order? This shouldn't be a concern of the user of the type as it's easy to forget or otherwise mess up.

ZhidongPeng replied:
This entry is get from c/ebpf directly, I can wrap this property into a function to mitigate this problem.

ZhidongPeng replied:
Added comments to these fields and functions to get the value properly.

jeremycline replied:
Remove the `pub` marker from the fields so users can't get a hold of them without going through the accessor function.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,689 @@
+    ip = redirector::ip_to_string(entry.destination_ipv4);
+    port = u16::from_be(entry.destination_port); // convert a 16-bit number from network byte order to host byte order
+    Connection::write(connection_id, format!("Use lookup value:{ip}:{port}."));
+    connection.ip = ip.to_string();
```
This `ip` variable is coming from a function called `ip_to_string`. I've not got this open in an IDE so I'm not sure what the actual type is, but either the function is mis-named, or this is unnecessary. In any case, I don't think you should store it in this struct as a string, keep it as a meaningful type.

ZhidongPeng replied:
redirector::ip_to_string is convert the ip v4 address in u32 format to string, I switched to use IPv4Addr::from_bits fun

ZhidongPeng replied:
I also converted the string type to IIv4Addr struct.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,689 @@
+        ip.to_string(),
+        port,
+        connection_id,
+        request.uri().to_string(),
```
I don't know how this function is using the URI, but it seems more appropriate to use the `Uri` type you get from the function call instead of turning it into a string.

ZhidongPeng replied:
request.uri() is relative Url here, it is hyper::Uri struct, while in our authenticate logic, we used another Url crate struct which support query_pairs(). 
I will convert the request.uri() to Url crate struct and pass down.

jeremycline replied:
[hyper::Uri](https://docs.rs/hyper/latest/hyper/struct.Uri.html) has scheme, host, and port components as well so your comment about it being relative doesn't make sense. It doesn't handle parsing the query, but `Url` just uses [form_urlencode](https://docs.rs/form_urlencoded/1.2.1/form_urlencoded/fn.parse.html) to provide that feature.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,689 @@
+    );
+    proxy_request.headers_mut().insert(
+        HeaderName::from_static(constants::CLAIMS_HEADER),
+        HeaderValue::from_str(&host_claims).unwrap(),
```
Replace the `unwrap()` here with proper error handling.

ZhidongPeng replied:
resolved, if failed to insert this header, it will response bad gateway error

ZhidongPeng replied:
done.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,689 @@
+    );
+    proxy_request.headers_mut().insert(
+        HeaderName::from_static(constants::DATE_HEADER),
+        HeaderValue::from_str(&misc_helpers::get_date_time_rfc1123_string()).unwrap(),
```
Replace the `unwrap()` here with proper error handling

ZhidongPeng replied:
resolved, if failed to insert this header, it will response bad gateway error

ZhidongPeng replied:
done.

jeremycline commented on proxy_agent/src/proxy/proxy_server.rs:
```diff
@@ -0,0 +1,689 @@
+        return Ok(empty_response(StatusCode::PAYLOAD_TOO_LARGE));
+    }
+
+    let whole_body = match body.collect().await {
```
What's to stop a client from lying in the `Content-Length` header? Does `collect()` on body take the `Content-Length` header into account in all cases (I suspect the answer is "no" and see no unit tests here to prove otherwise)?

As I previously commented, use a tried and tested implementation to limit body sizes rather than implementing it yourself. [Tower has middleware for this](https://docs.rs/tower-http/latest/tower_http/limit/index.html).

ZhidongPeng replied:
tower-http middleware controls all the requests, while in our case, it skips the collect() if the requests do not require signature. Like if VMAgent put VMAgent zip log file (up to 100MB) to host, it will stream the data to host without any buffer in GPA.

jeremycline replied:
I am reasonably confident that you can design things such that the middleware is only applied to a sub-set of requests. Axum is a web framework that uses Hyper and Tower for middleware and you can do this. If you can't figure it out I'm happy to look harder at the implementation details, but your current approach is almost certainly incorrect.

To be honest, you might be able to just use Axum directly, so you don't re-invent that particular wheel.

ZhidongPeng replied:
used use tower service as middleware to limit the request body size

jeremycline commented on proxy_agent/src/proxy/windows.rs:
```diff
@@ -111,9 +103,14 @@ fn load_users() -> HashMap<u64, &'static str> {
+
     unsafe {
-        let mut user_name = "undefined".to_string();
+        let mut user_name;
```
This change doesn't seem related to everything else.

As I mentioned elsewhere, this isn't not how to use cancellation tokens; they should be `select!`ed on with whatever future you want to let someone else cancel and the callee doesn't need to have any knowledge of it.

As an aside, this `unsafe` block needs to be replaced with minimal blocks around the actually-unsafe calls. Please do that in a separate PR, however.

ZhidongPeng replied:
tokio::select! is designed for asynchronous contexts or async fn, in my case, it is not async fn, I have to use cancellation_token.is_cancelled() to determine this function should move forward or return interrupted error earlier.

ZhidongPeng replied:
removed the large unsafe section, instead of, I added some unsafe to the statements only needed.

jeremycline replied:
Please back out all these changes. Fix the unsafe block in a separate PR as I asked, this one is already too big.

Regarding this function generally: you're using the `windows-sys` API, which is following the normal Rust convention of providing the unsafe binding to the FFI. There's a `windows` crate that provides a safe wrapper around that API. Is there no API in that crate (for example [this one](https://microsoft.github.io/windows-docs-rs/doc/windows/System/struct.User.html#method.GetFromId)) that does what you need? If there is no safe function, perhaps we can contribute one to the crate. 

This implementation is clearly wrong in at least one place: you call `let _status = Identity::LsaGetLogonSessionData(&luid, data.as_mut_ptr());` and then access data without checking the return value, but [it's a fallible call](https://learn.microsoft.com/en-us/windows/win32/api/ntsecapi/nf-ntsecapi-lsagetlogonsessiondata#return-value).

Finally, regarding cancellation: you don't use `select!` in this function, you would use it further up the call stack on a relatively long-running task you may want to cancel. It makes no sense to have a function that accepts a cancellation token, then immediately checks it and returns an error if it's cancelled. Why would the caller even call the function if it's already been cancelled? What's the run time of this function in the worst case? If it's relatively long, you shouldn't be calling it in any asynchronous context. If it's not relatively long, what's the value of cancelling it?

ZhidongPeng replied:
got it, I will remove the check_cancellation_token here and just keep the select! in the upper caller.

jeremycline commented on proxy_agent/src/service.rs:
```diff
@@ -4,20 +4,18 @@
 use std::time::Duration;
 
-pub fn start_service(shared_state: Arc<Mutex<SharedState>>) {
+pub async fn start_service(shared_state: Arc<Mutex<SharedState>>) {
```
As with several other functions in this PR, this function never awaits anything and does not need to be marked as `async.

ZhidongPeng replied:
you are right, the async is no longer needed after I moved to tokio::spawn instead of calling the async methods.

reviewed all the async keywords in the project and removed unnecessary ones.

jeremycline reviewed (commented):
Once again, this isn't a complete review, I'm time-boxing this and there's plenty of substantial changes to be made here.

ZhidongPeng committed:
use tokio::select! cancellation_token to break/reaturn the loop

ZhidongPeng committed:
Merge branch 'dev1' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

ZhidongPeng committed:
resolve comments.

ZhidongPeng committed:
make linux build

ZhidongPeng committed:
use tower middleware to limit the request body size.

ZhidongPeng committed:
Merge branch 'dev1' of https://github.com/ZhidongPeng/GuestProxyAgent into dev1

ZhidongPeng committed:
refactor the code to not re-create the connection context in handle_reqeust

jeremycline commented on proxy_agent/src/main.rs:
```diff
@@ -72,22 +86,31 @@ fn main() {
+                );
+            });
+
+            main.join().unwrap();
```
Ah, this will block one of the runtime threads which defeats the purpose of spawning the thread.

There's a couple options here, but the easiest is to (ab)use [spawn_block](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html). You're not _really_ meant to use it for things that last the length of the program, but the default pool size is 512 and you're not otherwise doing much heavy i/o that doesn't have async support.

ZhidongPeng replied:
I forget why I use the thread::spawn, it looks like I do not need to spawn and block it here.

jeremycline commented on proxy_agent/src/common/http.rs:
```diff
@@ -1,478 +1,456 @@
+    let (content_type, charset_type) =
+        if let Some(content_type) = response.headers().get(hyper::header::CONTENT_TYPE) {
+            if let Ok(content_type_str) = content_type.to_str() {
+                let content_type_str = content_type_str.to_lowercase();
+                let content_type;
+                if content_type_str.contains("xml") {
+                    content_type = "xml";
+                } else if content_type_str.contains("json") {
+                    content_type = "json";
+                } else if content_type_str.contains("text") {
+                    content_type = "text";
+                } else {
+                    content_type = "unknown";
+                }
 
-    Ok(lines)
-}
+                let charset_type;
+                if content_type_str.contains("utf-8") {
+                    charset_type = "utf-8";
+                } else if content_type_str.contains("utf-16") {
+                    charset_type = "utf-16";
+                } else if content_type_str.contains("utf-32") {
+                    charset_type = "utf-32";
+                } else {
+                    charset_type = "unknown";
+                }
```
There's a [mime](https://docs.rs/mime/latest/mime/) library you can use to parse things. It's already in your dependency tree via tower and it's very widely used.

ZhidongPeng replied:
Thanks!

jeremycline reviewed (commented):
Once again, I've not done a full review, nor do I expect that I will have the time to do so. This pull request is huge and continues to add unrelated changes despite my requests to not do so.

My recommendation would be to focus on cleaning up some of the foundations of this project before focusing on anything else. Get rid of the custom logging and use tracing. Clean up the error handling with `anyhow`. Reviewing this is difficult in part because the error handling and logging is cumbersome.

ZhidongPeng commented:
This big PR break down into 3 smaller ones.
